deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java

```diff
@@ -153,6 +159,75 @@
     private final PropertyChangeListener propertyChangeListener;
     private Future<?> undeployScanTask;
 
+    @Override
+    public void handleNotification(Notification notification) {
+        try {
+            if (acquireScanLock()) {
+                ROOT_LOGGER.warn("Notification processed");
+                switch (notification.getType()) {
+                    case DEPLOYMENT_DEPLOYED_NOTIFICATION: {
+                        String runtimeName = notification.getData().get(DEPLOYMENT).asString();
+                        if (!deployed.containsKey(runtimeName)) {
+                            updateDeploymentStatusAfterNotification(deploymentDir.toPath(), runtimeName);
+                        }
+                        break;
+                    }
+                    case DEPLOYMENT_UNDEPLOYED_NOTIFICATION: {
+                        String runtimeName = notification.getData().get(DEPLOYMENT).asString();
+                        if (deployed.containsKey(runtimeName)) {
+                            clearMarkers(deployed.get(runtimeName).parentFolder.toPath(), runtimeName);
+                            deployed.remove(runtimeName);
+                        }
+                        break;
+                    }
+                    default:
+                    //ignore
+                }
+            }
+        } finally {
+            releaseScanLock();
+        }
```

Comment:
The try/finally should be inside the if (acquireScanLock()) { block so you don't release something you hadn't acquired.